File: Week_08/java/src/_56_MergeIntervals.java
import java.util.Arrays;
import java.util.Comparator;
/**
 * 第一遍：2020/07/12周日 ✅
 * 第二遍：2020/07/13周一 ✅
 * 第二遍：2020/07/06周一
 * 第二遍：2020/07/03周五
 * 第三遍：2020/06/29周一
 * 第四遍：2020/07/05周日
 */
class _56_MergeIntervals {
    public int[][] merge(int[][] intervals) {
        if (intervals == null || intervals.length == 0) return new int[0][];
        Arrays.sort(intervals, Comparator.comparingInt(e -> e[0]));
        int length = intervals.length;
        int[][] ret = new int[length][2];
        ret[0] = intervals[0];
        int j = 0;

        for (int i = 1; i < length; i++) {
            int[] current = intervals[i];
            if (current[0] > ret[j][1]) {
                ret[++j] = current;
            } else {
                ret[j][1] = Math.max(ret[j][1], current[1]);
            }
        }
        return Arrays.copyOf(ret, j + 1);
    }
}

File: Week_08/cpp/src/_56_MergeIntervals.cpp
#include <vector>
using namespace std;
class Solution
{
public:
    // O(nlogn) O(1)
    vector<vector<int>> merge(vector<vector<int>> &intervals)
    {
        // 考虑某一个区间[l, r]与之重叠的所有区间分散在数组中的各个位置
        // 所以需要扫描数组中的每一个元素以找到与[l, r]重叠的所有区间
        // 因此暴力法是O(n^2)
        // 能否通过某种方式将不连续的重叠区间放在连续的位置上？
        // 通过将数组以区间的左端点排序可以做到
        // 这样先排序，然后顺序扫描排序后的数组顺序将每一个重叠的区间集合合并成一个区间，
        // 然后将该区间放到结果数组中
        if (intervals.empty())
            return {};
        sort(intervals.begin(), intervals.end());
        vector<vector<int>> ret;
        auto *p = &intervals[0];
        for (auto &it : intervals)
        {
            if ((*p)[1] < it[0])
            {
                ret.push_back(*p);
                p = &it;
            }
            else
            {
                (*p)[1] = max((*p)[1], it[1]);
            }
        }
        ret.push_back(*p);
        return ret;
    }
};